src/v/model/ktp.h

```diff
@@ -0,0 +1,301 @@
+     */
+    topic_partition_view as_tp_view() const {
+        return {get_topic(), get_partition()};
+    }
+
+    /**
+     * @brief Returns a reference to the ktp's topic
+     */
+    const model::topic& get_topic() const { return tp.topic; }
+
+    /**
+     * @brief Returns the ktp's partition id
+     */
+    model::partition_id get_partition() const { return tp.partition; }
+
+    /**
+     * @brief Returns the ktp's namespace as a string_view.
+     *
+     * Of course, the namespace is always "kafka", but this method
+     * is provided to ease interop with other types which may have
+     * variable namespaces.
+     */
+    constexpr const std::string_view get_ns() { return kafka_ns_view; }
+
+    /**
+     * @brief Compare with another ktp.
+     *
+     * This is more efficient than ntp-only or mixed ntp <-> ktp comparison
+     * since it does not need to compare the namespace at all.
+     */
+    bool operator==(const ktp& other) const { return tp_equals(tp, other.tp); }
+
+    /**
+     * @brief Compare with an ntp.
+     *
+     * Although this must compare the namespaces it is still in practice faster
+     * than an ntp <-> ntp comparison, because one side of the namespace
+     * comparison is constexpr, which leads to very efficient code generation
+     * (an inlined, fixed-size memcmp) for that comparison.
+     */
+    bool operator==(const ntp& other) const {
+        return tp_equals(other.tp, tp) && kafka_ns_view == other.ns();
+    }
+
+    friend std::ostream& operator<<(std::ostream& os, const ktp& t) {
+        os << t.as_ntp();
+        return os;
+    }
+
+private:
+    static bool tp_equals(
+      const model::topic_partition& l, const model::topic_partition& r) {
+        return l.partition == r.partition
+               && std::equal_to<std::string_view>{}(l.topic(), r.topic());
+    }
+};
+} // namespace model
+
+namespace std {
+template<>
+struct hash<model::ktp> {
+    size_t operator()(const model::ktp& ktp) const {
+        return detail::ntp_hash(
+          model::kafka_ns_view, ktp.get_topic()(), ktp.get_partition());
+    }
+};
+} // namespace std
+
+namespace model {
+
+/**
+ * @brief A ktp object that remembers its hash code.
+ *
+ * In some contexts, a ktp/ntp object will be repeatedly hashed, which
+ * is a relatively expensive operation. There, we can swap in this type
+ * which remembers its hash code. It is a subclass of ktp, and can safely
+ * be passed as a reference to a ktp or sliced down to a ktp for convenience.
+ *
+ * The hash code is calculated once at constructon and never changes, since
+ * the underlying topic and partition cannot be changed (but this invariant
+ * must be enforced by the ktp class as noted there). Among other things this
+ * means that assignment of a ktp_with_hash from a ktp object does not compile:
+ * for this to be safe we'd need to recalculate the hash after assignment.
+ */
+struct ktp_with_hash : public ktp {
```

Comment:
qq: why is the `ktp` base class not always `ktp_with_hash` is there a use case where we don't want the hash to be remembered? 

and if this is the case, wondering if we simply port this change to the `ntp` object altogether - at 30K partitions you are looking at 240kb. 

i feel like it could be something like a `alias fuzzy_id = named_typed<uint64_t>` ... just thoughts. wondering if you considered that design space of pushing the memoized hash code to the ntp base class. and potentially optimizing the storage of the ntp itself.

Reply:
> why is the ktp base class not always ktp_with_hash is there a use case where we don't want the hash to be remembered?

I think having only one `ktp` class with always-memorized hash is also a reasonable design and would definitely be simpler. `ktp` exists mostly because it's a "strictly better" `ntp` for the use case of kafka partitions. `ktp_with_hash` is worse than `ktp` in scenarios where you hash it zero or 1 times, since it is larger, has somewhat slower copy & move constructors/operators, etc. For example, using `ktp_with_hash` as a key in a hash_map may not make much sense: especially if the map itself stores the hashes (otherwise, the cached hash is at least useful during re-hashes).

> and if this is the case, wondering if we simply port this change to the ntp object altogether - at 30K partitions you are looking at 240kb.

In practice there might be many more ntp objects, because we hardly have just one copy of every ntp and in fact we probably have many copies. Though regardless it's probably a small relative increase when you consider that the total size of the ntp object is probably already > 60 bytes usually (when you count the size of the two embedded strings and their allocations, etc).

> i feel like it could be something like a alias fuzzy_id = named_typed<uint64_t> ... just thoughts. wondering if you considered that design space of pushing the memoized hash code to the ntp base class. and potentially optimizing the storage of the ntp itself.

This is targeted at a backport so I didn't want to get totally crazy. That said, the next step is something exactly like that, but not ntp specific: for _any_  object with type `T` where `T` is a regular type we should be able to "intern" it to create a handle<T> which would be only 8 bytes (essentially a pointer). It's not fuzzy, it's exact: two handles represent equal objects iff they are equal, so comparison is a single 8-byte compare, hashing can hash the 8-byte handle or be memorized, etc. This would be perfect for ntp and similar objects that we have a lot of. boost offers this as `boost::flyweight` but we need a seastar flavored version (in particular, the lookup must not block and yet handles should be global (i.e., the same object interned on two shards must result in the same handle).

That's too much for a patch tho.

Reply:
>boost offers this as boost::flyweight

thank you so much. the docs are good and fill in a lot of missing details that i was making up in my head about how this might work.

Reply:
@dotnwat - since I think this code falls solidly in the Enterprise team's boundaries, what do you think about separate ktp + ktp_with_hash classes versus just the one which always caches the hash code? I'm fine to go either way here.

Reply:
OK, I'll leave this as is here, agreed on the stepping-stone part.